File: javascript/src/iterator/value/GenericCollectionIteratorValue.ts
//··························································
// Copyright (c) 2023-2026. Jonathan Bédard ~ JóôòKiwi
//
// This project is free to use.
// All the right is reserved to the author of this project.
// My projects:
//  - https://github.com/joooKiwi/type
//  - https://github.com/joooKiwi/lazy
//  - https://github.com/joooKiwi/collection
//  - https://github.com/joooKiwi/enumeration
//··························································

import type {CollectionHolder}        from "../../CollectionHolder"
import type {CollectionIterator}      from "../CollectionIterator"
import type {CollectionIteratorValue} from "./CollectionIteratorValue"

import {AbstractIteratorValue} from "./AbstractIteratorValue"

/**
 * An implementation of a {@link CollectionIteratorValue}
 * with specified {@link CollectionIterator} (for debugging purpose),
 * a {@link CollectionHolder} (for debugging purpose)
 * and the {@link index} to give the index that the {@link value} needs to be retrieved.
 *
 * @typeParam T          The element type (`unknown` by default)
 * @typeParam REFERENCE  The {@link CollectionIterator} source (`CollectionIterator<T>` by default)
 * @typeParam COLLECTION The {@link CollectionHolder} source (`CollectionHolder<T>` by default)
 * @typeParam INDEX      The index to retrieve in the {@link collection} (`number` by default)
 */
export class GenericCollectionIteratorValue<const T = unknown,
    const REFERENCE extends CollectionIterator<T> = CollectionIterator<T>,
    const COLLECTION extends CollectionHolder<T> = CollectionHolder<T>,
    const INDEX extends number = number, >
    extends AbstractIteratorValue<T, false>
    implements CollectionIteratorValue<T> {

    readonly #iterator
    readonly #collection
    readonly #index

    public constructor(iterator: REFERENCE, collection: COLLECTION, index: INDEX,) {
        super()
        this.#iterator = iterator
        this.#collection = collection
        this.#index = index
    }

    /** The {@link CollectionIterator} that was the origin of the current {@link CollectionIteratorValue instance} */
    protected get _iterator(): REFERENCE { return this.#iterator }

    public get collection(): COLLECTION { return this.#collection }

    public get index(): INDEX { return this.#index }

    public override isDone(): false { return false }
    public override isNotDone(): true { return true }

    public override get value(): T { return this.collection.get(this.index,) }

}

File: java/collection.joookiwi/src/main/java/joookiwi/collection/java/iterator/value/GenericCollectionIteratorValue.java
package joookiwi.collection.java.iterator.value;

import joookiwi.collection.java.CollectionHolder;
import joookiwi.collection.java.iterator.CollectionIterator;
import org.jetbrains.annotations.Contract;
import org.jetbrains.annotations.NotNullByDefault;
import org.jetbrains.annotations.Nullable;
import org.jetbrains.annotations.Range;

import static joookiwi.collection.java.CommonContracts.ALWAYS_FALSE_0;
import static joookiwi.collection.java.CommonContracts.ALWAYS_TRUE_0;
import static joookiwi.collection.java.NumericConstants.MAX_INT_VALUE;

/// An implementation of a [CollectionIteratorValue]
/// with specified [CollectionIterator] (for debugging purpose),
/// a [CollectionHolder] (for debugging purpose)
/// and the [#index] to give the index that the [#value] needs to be retrieved
///
/// @param <T>          The element type
/// @param <COLLECTION> The [CollectionHolder] source
@NotNullByDefault
public class GenericCollectionIteratorValue<T extends @Nullable Object,
        REFERENCE extends CollectionIterator<? extends T>,
        COLLECTION extends CollectionHolder<? extends T>>
        extends AbstractIteratorValue<T>
        implements CollectionIteratorValue<T> {

    private final REFERENCE __iterator;
    private final COLLECTION __collection;
    private final @Range(from = 0, to = MAX_INT_VALUE) int __index;

    public GenericCollectionIteratorValue(final REFERENCE iterator, final COLLECTION collection, final @Range(from = 0, to = MAX_INT_VALUE) int index) {
        super();
        __iterator = iterator;
        __collection = collection;
        __index = index;
    }

    @Contract(pure = true)
    protected REFERENCE _iterator() { return __iterator; }

    @Contract(pure = true)
    @Override public COLLECTION collection() { return __collection; }

    @Contract(pure = true)
    @Override public @Range(from = 0, to = MAX_INT_VALUE) int index() { return __index; }

    @Contract(value = ALWAYS_FALSE_0, pure = true)
    @Override public boolean isDone() { return false; }

    @Contract(value = ALWAYS_TRUE_0, pure = true)
    @Override public boolean isNotDone() { return true; }

    @Override public T value() { return collection().get(index()); }

}
